tdms/tests/unit/test_interpolate.cpp

```diff
@@ -0,0 +1,128 @@
+# include "catch2/catch_test_macros.hpp"
+# include "interpolate.h"
+
+using namespace std;
+
+/**
+ * @brief Test whether checkInterpolationPoints throws exceptions in the correct cases.
+ * 
+ * This function checks whether it is possible to use cubic interpolation over a given index range.
+ */
+TEST_CASE("checkInterpolationPoints: exceptions thrown") {
+
+    // setup some fake field dimensions
+    int I = 6, J = 7, K = 8;
+
+    // 2D simulations have J = 0 - no width in this dimension
+    // checkInterpolationPoints should not throw an exception if we provide j_l = 2 > j_u = -2, as is done for a 2D simulation (J=0), and request interpolation over the maximum i,k range
+    CHECK_NOTHROW(checkInterpolationPoints(2, I - 2, 2, -2, 2, K - 2, I, 0, K));
+
+    // conversely, it should throw an error if we try to provide j_l = 0 = j_u, which one may percieve are sensible choices, in this case
```

Comment:
> which one may perceive are sensible choices 

That's exactly what I would perceive. Is the j_l = j_u != 0 requirement from the previous implementation?

Reply:
Yep - original code seems to rely on this behaviour, so I thought it best to preserve it for now

Reply:
@prmunro : is the correct behaviour? it seems a little odd

Reply:
@t-young31 , can you please direct me to the place in the original code where this behavior is encountered? Sorry if this should be obvious (in which case just tell me how I should have found it).

Reply:
@willGraham01 did you find where this was occurs in iterator.cpp ?

Reply:
An example of the original error check [is here](https://github.com/UCL/TDMS/blob/9a8a2291ec4783a320e1a9b75018d5e92bfb0082/tdms/src/interpolate.cpp#L144-L165) and use of the `k_l=0=k_u` case is [here](https://github.com/UCL/TDMS/blob/9a8a2291ec4783a320e1a9b75018d5e92bfb0082/tdms/src/iterator.cpp#L6071-L6073).
(The other TE/TM interpolation functions also have similar error checks, but I'll stick to talking about this specific instance as the others are similar.) 

It is not checked that `k_l < k_u`. However it looks like this would be taken care of implicitly, as the [`for` loops that follow](https://github.com/UCL/TDMS/blob/9a8a2291ec4783a320e1a9b75018d5e92bfb0082/tdms/src/interpolate.cpp#L168-L174) would be empty, so "nothing would be done". Whether this is the intended behaviour or not I wasn't sure, but I've preserved the behaviour by _not_ letting `checkInterpolationPoints` throw an error in the `k_l < k_u` case.

Reply:
I would have thought `k_l <= k_u` is definitely the intended behaviour. probably worth adding that as an additional constraint(?)

Reply:
Adding this check causes the tests to fail when running `run_pstd_bscan`; see https://github.com/UCL/TDMS/runs/7896933760?check_suite_focus=true.
At some point in the 2D simulation we need to allow j_l > j_u in order to not throw an error, because the interpolation functions rely on j_l = 2 > j_u = -2 being true.
For now, I've avoided adding this check until we're completely clear on what is relying on it. Also, the new `E-field` interpolation functions that act component-wise rather than mode-wise, won't depend on this behaviour.

Reply:
exciting
very happy to leave out that check if that possibility will be removed later anyway 👍🏼